src/libcore/ptr/mod.rs

```diff
@@ -1074,17 +1074,24 @@ impl<T: ?Sized> *const T {
     /// operation because the returned value could be pointing to invalid
     /// memory.
     ///
-    /// When calling this method, you have to ensure that if the pointer is
-    /// non-NULL, then it is properly aligned, dereferencable (for the whole
-    /// size of `T`) and points to an initialized instance of `T`. This applies
-    /// even if the result of this method is unused!
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or* 
+    /// all of the following is true:
+    /// - it is properly aligned
+    /// - it is dereferencable, i.e the memory range of the given `size_of::<T>()` 
+    ///   the pointer points to must all be within the bounds of a single allocated object 
+    ///   (possibly stack allocated) *and* this object must be an initialized instance of `T`
+    ///
+    /// This applies even if the result of this method is unused!
     /// (The part about being initialized is not yet fully decided, but until
     /// it is, the only safe approach is to ensure that they are indeed initialized.)
     ///
     /// Additionally, the lifetime `'a` returned is arbitrarily chosen and does
-    /// not necessarily reflect the actual lifetime of the data. It is up to the
-    /// caller to ensure that for the duration of this lifetime, the memory this
-    /// pointer points to does not get written to outside of `UnsafeCell<U>`.
+    /// not necessarily reflect the actual lifetime of the data. Moreover, you *must* enforce 
+    /// Rust aliasing rules, more specifically: for the duration of this lifetime, 
+    /// the memory the pointer points to does not get mutated (except inside `UnsafeCell`).
+    ///
+    /// Also, you must ensure that at any gimen moment no immutable reference is aliased 
+    /// with a mutable reference pointing to the same allocation object, *even if unused*. 
```

Comment:
I'm not sure how to phrase it properly: we need to account for existence of splitting borrows. Is "allocation object" here correct?

Reply:
cc @RalfJung any thoughts?

Reply:
I wouldn't try to define Rust's aliasing rules here.

Reply:
What I meant is, I think this paragraph is not useful. This is a general statement about Rust aliasing, where we are still figuring out the details, and I think the statement is both too vague and too detailed and at the wrong place.